File: proxypattern/proxypattern.js
class EmployeeDAO {
    create(client, employeeObj) {
        throw new Error('Abstract method "create()" must be implemented');
    }

    delete(client, employeeId) {
        throw new Error('Abstract method "delete()" must be implemented');
    }

    get(client, employeeId) {
        throw new Error('Abstract method "get()" must be implemented')
    }
}

class EmployeeDAOImpl extends EmployeeDAO {
    create(client, employeeObj) {
        // creates a new row
        console.log('created a new row in employee table')
    }

    delete(client, employeeId) {
        // delete a row
        console.log('deleted a row with employee id:', employeeId)
    }

    get(client, employeeId) {
        // fetch row 
        console.log('fetching data from the DB')
        return new EmployeeDO()
    }
}

class EmployeeDAOProxy extends EmployeeDAO {
    constructor() {
        super()
        this.employeeDAOObj = new EmployeeDAOImpl()
    }

    create(client, employeeObj) {
        if (client == 'ADMIN') {
            this.employeeDAOObj.create(client, employeeObj)
            return 
        } else {
            console.log('Access Denied!')
        }
    }

    delete(client, employeeId) {
        if (client == 'ADMIN') {
            this.employeeDAOObj.delete(client, employeeId)
            return 
        } else {
            console.log('Access Denied!')
        }
    }

    get(client, employeeId) {
        if (client == 'ADMIN') {
            this.employeeDAOObj.get(client, employeeId)
            return 
        } else {
            console.log('Access Denied!')
        }
    }
}

class EmployeeDO {
    constructor(name, id) {
        this.name = name
        this.id = id
    }
}

const employeeTableObj = new EmployeeDAOProxy()
employeeTableObj.create('USER', new EmployeeDO('John', '23'))
employeeTableObj.delete('ADMIN', new EmployeeDO('Doe', '54').id)

File: proxypattern/proxypattern.py
from abc import ABC, abstractmethod

class EmployeeDAO(ABC):
    @abstractmethod
    def create(self, client, employeeObj):
        pass

    @abstractmethod
    def delete(self, client, employeeId):
        pass

    @abstractmethod
    def get(self, client, employeeId):
        pass

class EmployeeDAOImpl(EmployeeDAO):
    def create(self, client, employeeObj):
        # creates a new row
        print('created a new row in employee table')

    def delete(self, client, employeeId):
        # delete a row
        print('deleted a row with employee id: ', employeeId)

    def get(self, client, employeeId):
        # fetch row
        print('fetching data from the DB')
        return EmployeeDO("Dummy", employeeId)

class EmployeeDAOProxy(EmployeeDAO):
    def __init__(self):
        self.employeeDAOObj = EmployeeDAOImpl()

    def create(self, client, employeeObj):
        if client == 'ADMIN':
            self.employeeDAOObj.create(client, employeeObj)
            return 
        
        raise Exception('Access Denied!')

    def delete(self, client, employeeId):
        if client == 'ADMIN':
            self.employeeDAOObj.delete(client, employeeId)
            return 
        
        raise Exception('Access Denied!')

    def get(self, client, employeeId):
        if client == 'ADMIN' or client == 'USER':
            return self.employeeDAOObj.get(client, employeeId)
        
        raise Exception('Access Denied!')

class EmployeeDO():
    def __init__(self, name, id):
        self.name = name
        self.id = id
    
class ProxyDesignPattern:
    employeeTableObj = EmployeeDAOProxy()
    employeeTableObj.create('USER', EmployeeDO('John', '23'))
    # employeeTableObj.delete('ADMIN', EmployeeDO('Doe', '54').id)
    print('operation executed successfully')



